File: exampleretrofit/src/test/java/foo/bar/example/asafretrofit/feature/fruit/StateBuilder.java
package foo.bar.example.asafretrofit.feature.fruit;

import org.mockito.ArgumentCaptor;

import co.early.asaf.core.callbacks.FailureCallbackWithPayload;
import co.early.asaf.core.callbacks.SuccessCallbackWithPayload;
import co.early.asaf.retrofit.CallProcessor;
import foo.bar.example.asafretrofit.api.fruits.FruitPojo;
import foo.bar.example.asafretrofit.message.UserMessage;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.doAnswer;

/**
 *
 */
public class StateBuilder {

    private CallProcessor<UserMessage> mockCallProcessor;

    StateBuilder(CallProcessor<UserMessage> mockCallProcessor) {
        this.mockCallProcessor = mockCallProcessor;
    }

    StateBuilder getFruitSuccess(final FruitPojo fruitPojo) {

        final ArgumentCaptor<SuccessCallbackWithPayload> callback = ArgumentCaptor.forClass(SuccessCallbackWithPayload.class);

        doAnswer(__ -> {
            callback.getValue().success(fruitPojo);
            return null;
        })

                .when(mockCallProcessor)
                .processCall(any(), any(), callback.capture(), any());

        return this;
    }

    StateBuilder getFruitFail(final UserMessage userMessage) {

        final ArgumentCaptor<FailureCallbackWithPayload> callback = ArgumentCaptor.forClass(FailureCallbackWithPayload.class);

        doAnswer(__ -> {
            callback.getValue().fail(userMessage);
            return null;
        })

                .when(mockCallProcessor)
                .processCall(any(), any(), any(), callback.capture());

        return this;
    }

}

File: example-kt-07apollo/src/test/java/foo/bar/example/foreapollokt/feature/fruit/StateBuilder.kt
package foo.bar.example.foreapollokt.feature.fruit

import co.early.fore.kt.retrofit.CallProcessor
import co.early.fore.kt.retrofit.Either
import co.early.fore.retrofit.MessageProvider
import foo.bar.example.foreapollokt.api.fruits.FruitPojo
import foo.bar.example.foreapollokt.message.UserMessage
import io.mockk.coEvery
import kotlinx.coroutines.CompletableDeferred
import retrofit2.Response

/**
 *
 */
class StateBuilder internal constructor(private val mockCallProcessor: CallProcessor<UserMessage>) {

    internal fun getFruitSuccess(fruitPojo: FruitPojo): StateBuilder {

        coEvery {
            mockCallProcessor.processCallAsync(
                any() as suspend () -> Response<List<FruitPojo>>
            )
        } returns CompletableDeferred(Either.right(listOf(fruitPojo)))

        return this
    }

    internal fun getFruitFail(userMessage: UserMessage): StateBuilder {

        coEvery {
            mockCallProcessor.processCallAwait(
                any() as Class<MessageProvider<UserMessage>>,
                any() as suspend () -> Response<List<FruitPojo>>
            )
        } returns Either.left(userMessage)

        return this
    }
}
